Detectors/EMCAL/reconstruction/src/CaloRawFitterStandard.cxx

```diff
@@ -0,0 +1,143 @@
+// or submit itself to any jurisdiction.
+
+/// \file CaloRawFitterStandard.cxx
+/// \author Hadi Hassan (hadi.hassan@cern.ch)
+
+#include "FairLogger.h"
+#include <random>
+
+// ROOT sytem
+#include "TMath.h"
+#include "TF1.h"
+#include "TGraph.h"
+
+#include "EMCALReconstruction/AltroDecoder.h"
+#include "EMCALReconstruction/CaloFitResults.h"
+#include "DataFormatsEMCAL/Constants.h"
+
+#include "EMCALReconstruction/CaloRawFitterStandard.h"
+
+using namespace o2::emcal;
+
+CaloRawFitterStandard::CaloRawFitterStandard() : CaloRawFitter("Chi Square ( Standard )", "Standard")
+{
+  mAlgo = FitAlgorithm::Standard;
+}
+
+double CaloRawFitterStandard::rawResponseFunction(double* x, double* par)
+{
+  double signal = 0.;
+  double tau = par[2];
+  double n = par[3];
+  double ped = par[4];
+  double xx = (x[0] - par[1] + tau) / tau;
+
+  if (xx <= 0)
+    signal = ped;
+  else
+    signal = ped + par[0] * TMath::Power(xx, n) * TMath::Exp(n * (1 - xx));
+
+  return signal;
+}
+
+CaloFitResults CaloRawFitterStandard::evaluate(const std::vector<AltroDecoder::Bunch>& bunchlist,
+                                               std::optional<unsigned int> altrocfg1, std::optional<unsigned int> altrocfg2)
+{
+  float time = 0;
+  float amp = 0;
+  float chi2 = 0;
+  int ndf = 0;
+  bool fitDone = kFALSE;
+
+  auto [nsamples, bunchIndex, ampEstimate,
+        maxADC, timeEstimate, pedEstimate, first, last] = preFitEvaluateSamples(bunchlist, altrocfg1, altrocfg2, mAmpCut);
+
+  if (ampEstimate >= mAmpCut) {
+    time = timeEstimate;
+    int timebinOffset = bunchlist.at(bunchIndex).getStartTime() - (bunchlist.at(bunchIndex).getBunchLength() - 1);
+    amp = ampEstimate;
+
+    if (nsamples > 1 && maxADC < constants::OVERFLOWCUT) {
+      auto [amp, time, chi2, fitDone] = fitRaw(first, last);
+      time += timebinOffset;
+      timeEstimate += timebinOffset;
+      ndf = nsamples - 2;
+    }
+  }
+  if (fitDone) {
+    float ampAsymm = (amp - ampEstimate) / (amp + ampEstimate);
+    float timeDiff = time - timeEstimate;
+
+    if ((TMath::Abs(ampAsymm) > 0.1) || (TMath::Abs(timeDiff) > 2)) {
+      amp = ampEstimate;
+      time = timeEstimate;
+      fitDone = kFALSE;
+    }
+  }
+  if (amp >= mAmpCut) {
+    if (!fitDone) {
+      std::default_random_engine generator;
+      std::uniform_real_distribution<float> distribution(0.0, 1.0);
+      amp += (0.5 - distribution(generator));
+    }
+    time = time * constants::EMCAL_TIMESAMPLE;
+    time -= mL1Phase;
+
+    return CaloFitResults(-99, -99, mAlgo, amp, time,
+                          (int)time, chi2, ndf);
+  }
+  return CaloFitResults(-1, -1);
+}
+
+std::tuple<float, float, float, bool> CaloRawFitterStandard::fitRaw(int firstTimeBin, int lastTimeBin) const
+{
+
+  float amp(0), time(0), chi2(0);
+  bool fitDone(false);
+
+  int nsamples = lastTimeBin - firstTimeBin + 1;
+  fitDone = kFALSE;
+  if (nsamples < 3)
+    return std::make_tuple(amp, time, chi2, fitDone);
+
+  TGraph gSig(nsamples);
```

Comment:
In general using TGraph::Fit (i.e. the Minuit) is not a good idea for the reconstruction code, one should think about dedicated fitter (perhaps in different PR).

Reply:
We are well aware of the poor performance of the minuit-based raw fitters, which has manifested already in run2 on the HLT, for which we are looking into other fit methods (i.e. neural-network based). However this  is a porting of the standard method which was used in run2. Recently there was a commit of a prototype raw fitter to aliroot (https://github.com/alisw/AliRoot/pull/1054), not using minuit, which showed in first tests significant speedup (roughly factor 50) compared to the current one. Once this is fully commissioned we will port it to o2 and make it the default. As other chains (reco workflow, QC) depend on the raw fitter we would like to ask to port this raw fitter for the moment so that work in the other parts can continue.